### PR TITLE
Expand and improve AST expression support in Parquet readers

### PR DESCRIPTION
**Pull request by mhaseeb123**
## Description

Contributes to #17142
Supersedes #21489

This PR enables the Parquet readers to accept AST expressions of any form and passing them through predicate pushdown as no-op except the expression of the form `col op lit`, `lit op col`, `IS_NULL(col)`, `op(expr)` and `expr op expr`.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/parquet/experimental/dictionary_page_filter.cu [hunk 1]
```diff
@@ -1393,35 +1393,51 @@ class dictionary_expression_converter : public equality_literals_collector {
...
-        "Second operand of binary operation with column reference must be a literal");
-      v->accept(*this);
-
-      // Propagate the `_always_true` as expression to its unary operator parent
-      if (operator_arity == 1) {
+    using parquet::detail::extract_binary_operands;
+    using parquet::detail::extract_unary_operand;
+    using parquet::detail::operand_kind;
+
+    auto const input_op       = expr.get_operator();
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/experimental/dictionary_page_filter.cu:1399]: Separated out the same logic as before for unary and binary operators. Then for each operator, separated out the logic to handle different possible expression forms. e.g., `op col` for unary operators, `col op lit`, `lit op col`, `expr op expr` for binary operators and so on

File: cpp/src/io/parquet/bloom_filter_reader.cu [hunk 2]
```diff
@@ -199,59 +199,66 @@ class bloom_filter_expression_converter : public equality_literals_collector {
...
-      CUDF_EXPECTS(operator_arity == 1 or cudf::ast::detail::ast_operator_arity(op) == 2,
-                   "Only unary and binary operations are supported on column reference");
-      CUDF_EXPECTS(
-        operator_arity == 1 or dynamic_cast<ast::literal const*>(&operands[1].get()) != nullptr,
-        "Second operand of binary operation with column reference must be a literal");
-      v->accept(*this);
-
-      // Propagate the `_always_true` as expression to its unary operator parent
-      if (operator_arity == 1) {
+
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/bloom_filter_reader.cu:202]: > Separated out the same logic as before for unary and binary operators. Then for each operator, separated out the logic to handle different possible expression forms. e.g., `op col` for unary operators, `col op lit`, `lit op col`, `expr op expr` for binary operators and so on

Same thing as above. Just using the existing logic from lhs of this file

File: cpp/src/io/parquet/expression_transform_helpers.cpp [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+            .col_ref  = (lhs_kind == operand_kind::COLUMN_REF)
+                          ? dynamic_cast<ast::column_reference const*>(&lhs)
+                          : dynamic_cast<ast::column_reference const*>(&rhs),
+            .literal  = (rhs_kind == operand_kind::LITERAL)
+                          ? dynamic_cast<ast::literal const*>(&rhs)
+                          : dynamic_cast<ast::literal const*>(&lhs)};
+  }
+}
+
+named_to_reference_converter::named_to_reference_converter(
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/expression_transform_helpers.cpp:86]: The contents starting here until the eof are moved as is from `predicate_pushdown.cpp`

File: cpp/src/io/parquet/expression_transform_helpers.hpp [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+ *
+ * When the expression is of the form `lit op col`, the operator is inverted and the result
+ * is normalized so that col_ref and literal are set as if the form were `col op lit`.
+ */
+[[nodiscard]] binary_operands extract_binary_operands(ast::operation const& expr);
+
+/**
+ * @brief Collects column names from the expression ignoring the `skip_names`
+ */
+class names_from_expression : public ast::detail::expression_transformer {
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/expression_transform_helpers.hpp:162]: Starting this line until the EOF, the contents are moved here from `reader_impl_helpers.hpp` as is.

File: cpp/src/io/parquet/predicate_pushdown.cpp [hunk 5]
```diff
@@ -294,251 +293,4 @@ aggregate_reader_metadata::filter_row_groups(
      std::make_optional(num_bloom_filtered_row_groups)}};
 }
 
-/**
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/predicate_pushdown.cpp:297]: Moved as is to `expression_transform_helpers.cpp` as is

File: cpp/src/io/parquet/reader_impl_helpers.cpp [hunk 6]
```diff
@@ -1979,4 +1979,31 @@ aggregate_reader_metadata::select_columns(
     std::move(input_columns), std::move(output_columns), std::move(output_column_schemas));
 }
 
+std::vector<Type> aggregate_reader_metadata::get_parquet_types(
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/reader_impl_helpers.cpp:1985]: Moved from `bloom_filter_reader.cu` as is since it's also used in `dictionary_page_filter.cu`

File: cpp/src/io/parquet/reader_impl_helpers.hpp [hunk 7]
```diff
@@ -598,191 +594,4 @@ class aggregate_reader_metadata {
                  type_id timestamp_type_id);
 };
 
-/**
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/reader_impl_helpers.hpp:604]: Contents moved to `expression_transform_helpers.hpp` as is

File: cpp/src/io/parquet/stats_filter_helpers.cpp [hunk 8]
```diff
@@ -48,36 +48,39 @@ std::reference_wrapper<ast::expression const> stats_columns_collector::visit(
...
-      operator_arity == 1 or dynamic_cast<ast::literal const*>(&operands[1].get()) != nullptr,
-      "Second operand of binary operation with column reference must be a literal");
-    v->accept(*this);
-
-    // Return early if this is a unary operation
-    if (operator_arity == 1 and op != ast_operator::IS_NULL) { return expr; }
-
-    // Else if this is a supported binary operation, mark the column as needed
+
+  auto const input_op       = expr.get_operator();
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/stats_filter_helpers.cpp:52]: > Separated out the same logic as before for unary and binary operators. Then for each operator, separated out the logic to handle different possible expression forms. e.g., `op col` for unary operators, `col op lit`, `lit op col`, `expr op expr` for binary operators and so on

Same thing as above. Just using the existing logic from lhs of this file

File: cpp/tests/io/experimental/hybrid_scan_filters_test.cpp [hunk 9]
```diff
@@ -383,6 +370,79 @@ TEST_F(HybridScanFiltersTest, FilterRowGroupsWithStats)
   EXPECT_EQ(reader->total_rows_in_row_groups(stats_filtered_row_groups), 0);
 }
 
+TEST_F(HybridScanFiltersTest, FilterRowGroupsWithComplexExpressions)
```
Comment (mhaseeb123) [inline cpp/tests/io/experimental/hybrid_scan_filters_test.cpp:373]: New test added

File: cpp/tests/io/experimental/hybrid_scan_filters_test.cpp [hunk 10]
```diff
@@ -29,50 +29,35 @@ namespace {
...
-  auto const page_index_byte_range = reader->page_index_byte_range();
-
-  // Fetch page index bytes from the input buffer
-  auto const page_index_buffer =
-    cudf::io::parquet::fetch_page_index_to_host(datasource, page_index_byte_range);
-
-  // Setup page index
-  reader->setup_page_index(cudf::host_span<uint8_t const>{
-    static_cast<uint8_t const*>(page_index_buffer->data()), page_index_buffer->size()});
+  reader.reset_column_selection();
```
Comment (mhaseeb123) [inline cpp/tests/io/experimental/hybrid_scan_filters_test.cpp:50]: No need to create a new reader every time in this helper like lhs. Instead take in a pre-existing reader and just reset column selection.

File: cpp/tests/io/experimental/hybrid_scan_filters_test.cpp [hunk 11]
```diff
@@ -207,11 +192,12 @@ TEST_F(HybridScanFiltersTest, TestExternalMetadata)
...
-  using T                = cudf::timestamp_ms;
-  auto literal_value     = cudf::timestamp_scalar<T>(T(typename T::duration(100)));
-  auto literal           = cudf::ast::literal(literal_value);
-  auto col_ref_0         = cudf::ast::column_name_reference("col0");
-  auto filter_expression = cudf::ast::operation(cudf::ast::ast_operator::LESS, col_ref_0, literal);
+  using T            = cudf::timestamp_ms;
+  auto literal_value = cudf::timestamp_scalar<T>(T(typename T::duration(100)));
+  auto literal       = cudf::ast::literal(literal_value);
+  auto col_ref_0     = cudf::ast::column_name_reference("col0");
+  auto filter_expression =
```
Comment (mhaseeb123) [inline cpp/tests/io/experimental/hybrid_scan_filters_test.cpp:199]: Swap expression from `col0 < lit` to `lit > col0`

File: cpp/tests/io/experimental/hybrid_scan_filters_test.cpp [hunk 12]
```diff
@@ -337,10 +323,11 @@ TEST_F(HybridScanFiltersTest, FilterRowGroupsWithStats)
...
 
-  auto literal_value2     = cudf::string_scalar("000010000");
-  auto literal2           = cudf::ast::literal(literal_value2);
-  auto col_ref2           = cudf::ast::column_reference(2);
-  auto filter_expression2 = cudf::ast::operation(cudf::ast::ast_operator::LESS, col_ref2, literal2);
+  auto literal_value2 = cudf::string_scalar("000010000");
+  auto literal2       = cudf::ast::literal(literal_value2);
+  auto col_ref2       = cudf::ast::column_reference(2);
+  auto filter_expression2 =
+    cudf::ast::operation(cudf::ast::ast_operator::GREATER, literal2, col_ref2);
```
Comment (mhaseeb123) [inline cpp/tests/io/experimental/hybrid_scan_filters_test.cpp:330]: Swap expression from `col2 < lit2` to `lit2 > col2`

File: cpp/tests/io/experimental/hybrid_scan_filters_test.cpp [hunk 13]
```diff
@@ -465,7 +517,7 @@ TEST_F(HybridScanFiltersTest, FilterColumnSelection)
     auto filter_expression1 =
       cudf::ast::operation(cudf::ast::ast_operator::LESS, col_name0, literal1);
     auto filter_expression2 =
-      cudf::ast::operation(cudf::ast::ast_operator::LESS, col_ref2, literal2);
+      cudf::ast::operation(cudf::ast::ast_operator::GREATER, literal2, col_ref2);
```
Comment (mhaseeb123) [inline cpp/tests/io/experimental/hybrid_scan_filters_test.cpp:595]: Swap order and operator to test `lit op col` expression form

File: cpp/tests/io/experimental/hybrid_scan_filters_test.cpp [hunk 14]
```diff
@@ -645,10 +695,11 @@ TYPED_TEST(PageFilteringWithPageIndexStats, FilterPagesWithPageIndexStats)
...
   {
-    auto literal_value     = cudf::numeric_scalar<T>(T{100});
-    auto const literal     = cudf::ast::literal(literal_value);
-    auto const col_ref     = cudf::ast::column_name_reference("col0");
-    auto filter_expression = cudf::ast::operation(cudf::ast::ast_operator::LESS, col_ref, literal);
+    auto literal_value = cudf::numeric_scalar<T>(T{100});
+    auto const literal = cudf::ast::literal(literal_value);
+    auto const col_ref = cudf::ast::column_name_reference("col0");
+    auto filter_expression =
+      cudf::ast::operation(cudf::ast::ast_operator::GREATER, literal, col_ref);
```
Comment (mhaseeb123) [inline cpp/tests/io/experimental/hybrid_scan_filters_test.cpp:777]: Swap order and operator to test `lit op col` expression form

File: cpp/tests/io/parquet_reader_test.cpp [hunk 15]
```diff
@@ -1402,10 +1402,11 @@ TEST_F(ParquetReaderTest, FilterSimple)
...
   // Filtering AST - table[0] < RAND_MAX/2
-  auto literal_value     = cudf::numeric_scalar<decltype(RAND_MAX)>(RAND_MAX / 2);
-  auto literal           = cudf::ast::literal(literal_value);
-  auto col_ref_0         = cudf::ast::column_reference(0);
-  auto filter_expression = cudf::ast::operation(cudf::ast::ast_operator::LESS, col_ref_0, literal);
+  auto literal_value = cudf::numeric_scalar<decltype(RAND_MAX)>(RAND_MAX / 2);
+  auto literal       = cudf::ast::literal(literal_value);
+  auto col_ref_0     = cudf::ast::column_reference(0);
+  auto filter_expression =
+    cudf::ast::operation(cudf::ast::ast_operator::GREATER, literal, col_ref_0);
```
Comment (mhaseeb123) [inline cpp/tests/io/parquet_reader_test.cpp:1410]: Swap order and operator to test `lit op col` expression form

File: cpp/tests/io/parquet_reader_test.cpp [hunk 16]
```diff
@@ -1561,6 +1563,91 @@ TEST_F(ParquetReaderTest, FilterReferenceExpression)
   CUDF_TEST_EXPECT_TABLES_EQUAL(*result.tbl, *expected);
 }
 
+TEST_F(ParquetReaderTest, FilterColumnToColumnExpression)
```
Comment (mhaseeb123) [inline cpp/tests/io/parquet_reader_test.cpp]: New test for extended AST filter expressions

File: python/pylibcudf/tests/io/test_parquet.py [hunk 17]
```diff
@@ -100,16 +100,16 @@ def test_read_parquet_filters_metadata(tmp_path, if_prune_rowgroup, result):
...
+            ASTOperator.LESS,
             Literal(plc.Scalar.from_arrow(pa.scalar(max_element))),
+            ColumnNameReference("a"),
         )
     else:
         # No real pruning
         filter = Operation(
-            ASTOperator.GREATER,
-            ColumnNameReference("a"),
+            ASTOperator.LESS,
```
Comment (mhaseeb123) [inline python/pylibcudf/tests/io/test_parquet.py:110]: Swap order and operator to test `lit op col` expression form

**Comment by Matt711** [comment]
Thanks, testing this PR now...

File: cpp/src/io/parquet/stats_filter_helpers.cpp [hunk 18]
```diff
@@ -195,20 +213,14 @@ std::reference_wrapper<ast::expression const> stats_expression_converter::visit(
       }
       default: CUDF_FAIL("Unsupported binary operation in Statistics AST");
```
Comment (Matt711) [inline cpp/src/io/parquet/stats_filter_helpers.cpp]: For an expression like `pl.col("a").ne_missing(pl.lit(None, dtype=pl.Int64))` (i.e. NULL_EQUALS).
```suggestion
      _stats_expr.push(ast::operation{ast_operator::IDENTITY, *_always_true});
      return *_always_true;
```

**Comment by Matt711** [comment]
FYI Blocked by https://github.com/rapidsai/cudf/pull/21559

File: cpp/tests/io/experimental/hybrid_scan_filters_test.cpp [hunk 19]
```diff
@@ -207,11 +192,12 @@ TEST_F(HybridScanFiltersTest, TestExternalMetadata)
   }();
 
   // Filtering AST - table[0] < 100
```
Comment (vuule) [inline cpp/tests/io/experimental/hybrid_scan_filters_test.cpp]: does this need to be updated?

File: cpp/tests/io/experimental/hybrid_scan_filters_test.cpp (hunk 11 above)
Comment (vuule) [reply cpp/tests/io/experimental/hybrid_scan_filters_test.cpp:199]: Is this done because most tests cover col0 op lit?

File: cpp/tests/io/experimental/hybrid_scan_filters_test.cpp (hunk 19 above)
Comment (mhaseeb123) [reply cpp/tests/io/experimental/hybrid_scan_filters_test.cpp]: Oh yeah, it would now be `100 > table[0]`. Thanks for catching it

File: cpp/tests/io/experimental/hybrid_scan_filters_test.cpp (hunk 11 above)
Comment (mhaseeb123) [reply cpp/tests/io/experimental/hybrid_scan_filters_test.cpp:199]: Previously, we used to only support `col op lit` so all tests were in that format. I just swapped some of them to test the `lit op col` version here.

File: cpp/tests/io/experimental/hybrid_scan_filters_test.cpp
Comment (vuule) [inline cpp/tests/io/experimental/hybrid_scan_filters_test.cpp:1]: AFAICT there are no tests that check that complex expressions are correctly split so that row group pruning is still done as expected. Something like: "(col0 < 50) AND (col0 < col1)" should prune the same row groups as "(col0 < 50)".

Comment (mhaseeb123) [reply cpp/tests/io/experimental/hybrid_scan_filters_test.cpp:1]: We have this test in `parquet_reader_test.cpp`. Claude copied them over here too but since the internals for both APIs are the same, I removed it from here.

File: cpp/tests/io/parquet_reader_test.cpp [hunk 20]
```diff
@@ -1562,6 +1564,169 @@ TEST_F(ParquetReaderTest, FilterReferenceExpression)
...
+    auto predicate = cudf::compute_column(written_table, filter);
+    auto expected  = cudf::apply_boolean_mask(written_table, *predicate);
+
+    cudf::io::parquet_reader_options read_opts =
+      cudf::io::parquet_reader_options::builder(cudf::io::source_info{filepath}).filter(filter);
+    auto result = cudf::io::read_parquet(read_opts);
+    CUDF_TEST_EXPECT_TABLES_EQUAL(*result.tbl, *expected);
+  }
+
+  // Filter: (col_a < 150) or (col_a < col_b)
```
Comment (mhaseeb123) [inline cpp/tests/io/parquet_reader_test.cpp:1626]: This one

File: cpp/tests/io/experimental/hybrid_scan_filters_test.cpp
Comment (vuule) [reply cpp/tests/io/experimental/hybrid_scan_filters_test.cpp:1]: I don't think this is what I have in mind. I'm thinking of a test that uses hybrid scan so we can observe that with "(col0 < 50) AND (col0 < col1)" filtering actually happens in both stages - "(col0 < 50)" at row group level and "(col0 < col1)" at row level.

Comment (mhaseeb123) [reply cpp/tests/io/experimental/hybrid_scan_filters_test.cpp:1]: I guess the: 

```cpp
TEST_F(HybridScanFiltersTest, FilterRowGroupsWithComplexExpressions)
```

and the 

```cpp
TEST_F(HybridScanTest, ExtendedFilterExpressions)
```

do this. Just not both checks (at rg and row level) at the same time.

**Comment by ttnghia** [comment]
I think we should promote code review using LLMs. I ran it on this PR, still need @mhaseeb123 to verify:
```
---
  MUST FIX

  1. Missing return in bloom_filter_expression_converter::visit() -- unary fallthrough to binary code

  File: cpp/src/io/parquet/bloom_filter_reader.cu:207-230

  // Unary operation
  if (operator_arity == 1) {
      auto const [kind, col_ref] = extract_unary_operand(expr);
      if (kind == operand_kind::COLUMN_REF) {
          col_ref->accept(*this);
          _bloom_filter_expr.push(ast::operation{ast_operator::IDENTITY, *_always_true});
          return *_always_true;       // <-- returns OK
      } else {
          auto new_operands = visit_operands(expr.get_operands());
          if (&new_operands.front().get() == _always_true.get()) {
              _bloom_filter_expr.push(...);
              return *_always_true;   // <-- returns OK
          } else {
              _bloom_filter_expr.push(ast::operation{input_op, new_operands.front()});
              // <-- NO RETURN! Falls through to line 230
          }
      }
  }
  // Binary operation
  auto const [op, lhs_kind, rhs_kind, col_ref, literal] = extract_binary_operands(expr);
  //  ^^^ CUDF_EXPECTS(operands.size() == 2, ...) will throw for unary expr

  When a unary operation like NOT(col == 50) is processed and the child resolves to a non-_always_true
  expression (e.g., a bloom filter lookup), the code pushes the unary operation but does not return.
  Execution falls through to extract_binary_operands() on line 230, which calls
  CUDF_EXPECTS(operands.size() == 2, ...) -- this will throw at runtime.

  Comparison with the other two converters:
  - stats_expression_converter (stats_filter_helpers.cpp:154-155): has return _stats_expr.back(); after
   the unary else block
  - dictionary_expression_converter (dictionary_page_filter.cu:1420-1421): has return
  _dictionary_expr.back(); in the else branch

  Fix: Add return _bloom_filter_expr.back(); after line 224 (the else branch's push), matching the
  pattern in the other two converters. Also add a test for NOT(col == val) that exercises bloom filter
  filtering specifically.

  ---
  SHOULD FIX

  2. Typo in test comment

  File: cpp/tests/io/experimental/hybrid_scan_filters_test.cpp:410

  // Stats filter will prune based on col0 < 50 but pass throughcol0 < col1

  Missing space: "pass throughcol0" -> "pass through col0".

  3. binary_operands struct doc comments are slightly misleading

  File: cpp/src/io/parquet/expression_transform_helpers.hpp:49-50

  ast::column_reference const* col_ref;  ///< Non-null when exactly one operand is COLUMN_REF
  ast::literal const* literal;           ///< Non-null when exactly one operand is LITERAL

  For col op col, col_ref is set to one of the two column references while the other is lost, and
  literal will be nullptr. For expr op expr, both are nullptr. The comments should clarify: these are
  only reliably set for the col op lit / lit op col pattern; callers should check lhs_type/rhs_type
  before using them.

  4. names_from_expression::visit(column_reference) uses operator[] instead of .at()

  File: cpp/src/io/parquet/expression_transform_helpers.cpp:194

  auto const col_name = _column_indices_to_names[expr.get_column_index()];

  operator[] on std::unordered_map silently inserts a default (empty string) for missing keys. If a
  column reference has an index not in the map, this creates a phantom entry rather than failing fast.
  Should use .at() for bounds checking, consistent with how
  named_to_reference_converter::visit(column_name_reference) (line 134) uses
  _column_name_to_index.find() with CUDF_FAIL.

  ---
  SUGGESTIONS

  5. Add direct unit tests for extract_binary_operands normalization

  The function handles lit op col -> col op lit with operator inversion. Direct tests verifying:
  - lit < col normalizes to col > lit
  - col < lit stays as-is
  - col < col returns both kinds as COLUMN_REF
  - Commutative ops (==, !=) don't get inverted

  ...would be valuable for catching regressions in this shared infrastructure.

  6. Document the classify_operand assumption about column_name_reference

  File: cpp/src/io/parquet/expression_transform_helpers.cpp:46-55

  classify_operand classifies column_name_reference as EXPRESSION since it doesn't match
  column_reference or literal. This works because by the time expressions reach the filter converters,
  all name references should already be converted to index references. Worth adding a brief comment
  documenting this invariant.

  7. Consider factoring out the _always_true propagation pattern

  The pattern of checking _always_true and wrapping in IDENTITY appears in stats, bloom filter, and
  dictionary converters. While the different backing trees (_stats_expr, _bloom_filter_expr,
  _dictionary_expr) make extraction non-trivial, it could reduce the risk of inconsistencies like
  finding #1.

  ---
  Overall Assessment

  Well-structured and well-tested PR that significantly improves the flexibility of the Parquet
  reader's predicate pushdown. The refactoring into shared expression_transform_helpers is clean and
  eliminates real duplication. Test coverage is comprehensive, covering col op col, partial pushdown
  with AND/OR, NOT(col op lit), lit op col normalization, and complex nested expressions.

  One critical bug needs fixing: the bloom filter converter's missing return for unary operations
  (finding #1) will crash on expressions like NOT(col == 50) during bloom filter evaluation. The fix is
   straightforward (one line).
```

File: cpp/tests/io/experimental/hybrid_scan_filters_test.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/tests/io/experimental/hybrid_scan_filters_test.cpp:50]: It's a const API. Will be fine

**Comment by mhaseeb123** [comment]
Addressed all reviews from Claude code and Copilot

File: cpp/tests/io/parquet_reader_test.cpp [hunk 21]
```diff
@@ -1524,32 +1525,34 @@ TEST_F(ParquetReaderTest, FilterWithColumnProjection)
...
                          .filter(read_ref_expr);
-      EXPECT_THROW(cudf::io::read_parquet(read_opts), cudf::logic_error);
+      EXPECT_ANY_THROW(cudf::io::read_parquet(read_opts));
 
       // Repeat but select columns using indices instead of names
       read_opts = cudf::io::parquet_reader_options::builder(cudf::io::source_info{filepath})
                     .column_indices({2, 0})
                     .filter(read_ref_expr);
-      EXPECT_THROW(cudf::io::read_parquet(read_opts), cudf::logic_error);
+      EXPECT_ANY_THROW(cudf::io::read_parquet(read_opts));
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/tests/io/parquet_reader_test.cpp]: Fixed

File: cpp/src/io/parquet/expression_transform_helpers.cpp [hunk 22]
```diff
@@ -0,0 +1,347 @@
...
+}
+
+std::reference_wrapper<ast::expression const> names_from_expression::visit(
+  ast::column_reference const& expr)
+{
+  // Map the column index to its name
+  auto const col_name = _column_indices_to_names.at(expr.get_column_index());
+  // If the column name is not in the skip_names, add it to the set
+  if (_skip_names.count(col_name) == 0) { _column_names.insert(col_name); }
+  return expr;
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/io/parquet/expression_transform_helpers.cpp:183]: Fixed

File: cpp/src/io/parquet/stats_filter_helpers.cpp [hunk 23]
```diff
@@ -141,15 +141,34 @@ std::reference_wrapper<ast::expression const> stats_expression_converter::visit(
...
+    } else {
+      // For all other expression forms, visit operand and push expression
+      auto new_operands = visit_operands(expr.get_operands());
+      if (&new_operands.front().get() == _always_true.get()) {
+        // Pass through the _always_true child operand as is
+        _stats_expr.push(ast::operation{ast_operator::IDENTITY, _stats_expr.back()});
+        return *_always_true;
+      }
+      _stats_expr.push(ast::operation{input_op, new_operands.front()});
+      return _stats_expr.back();
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/io/parquet/stats_filter_helpers.cpp]: Good catch. Handled in 0f48f36

File: cpp/src/io/parquet/bloom_filter_reader.cu [hunk 24]
```diff
@@ -199,59 +199,51 @@ class bloom_filter_expression_converter : public equality_literals_collector {
...
-        }
-      }
+    }  // Visit operands and push expression for `expr op expr` form
+    else if (lhs_kind == operand_kind::EXPRESSION and rhs_kind == operand_kind::EXPRESSION) {
+      auto new_operands = visit_operands(expr.get_operands());
+      _bloom_filter_expr.push(ast::operation{op, new_operands.front(), new_operands.back()});
+    }  // Push _always_true for `col op col`, `expr op col`, `expr op lit` forms
+    else {
+      _bloom_filter_expr.push(ast::operation{ast_operator::IDENTITY, *_always_true});
     }
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/io/parquet/bloom_filter_reader.cu:247]: Done!

**Comment by ttnghia** [comment]
New review:
```
  MUST FIX (2)                                                                                                         

  1. Comment typo (stats_filter_helpers.cpp:156): Says "safely NULL it" but means "safely NOT/negate it"
  2. NOT(IS_NULL(expr)) edge case (stats_filter_helpers.cpp:157-165): When IS_NULL's operand is a sub-expression (not a
  column), the code pushes NOT(original_IS_NULL_ref) into _stats_expr, referencing the input expression tree rather than
  a stats-transformed expression. Should degrade to _always_true for NOT(IS_NULL(non_column)).

  SHOULD FIX (3)

  3. Missing [[nodiscard]] on function definitions in expression_transform_helpers.cpp — inconsistent with declarations
  in the header and other functions in the same file
  4. Inconsistent error handling: CUDF_FAIL in expression_transform_helpers.cpp:145 vs CUDF_EXPECTS for the same check in
   hybrid_scan_helpers.cpp:416
  5. classify_operand uses CUDF_FAIL for column_name_reference — should be unreachable if converter ran first; consider
  CUDF_UNREACHABLE or documenting as defensive

  SUGGESTIONS (5)

  6. get_parquet_types bug fix (using row_group_indices[src_index].front() instead of hardcoded 0) deserves a comment
  explaining why
  7. Initialize col_ref/literal to nullptr in extract_binary_operands for non-matching patterns
  8. apply_unary_membership_transform is a 30-line template in a header — acceptable for now but worth monitoring
  9. expr op col / col op expr branches correctly skip operand visiting since column collection is a separate pass —
  verified correct
  10. Test naming cleanup is good; remaining Test prefixes in untouched tests could be cleaned up separately

  Overall

  Solid refactoring with good code deduplication and comprehensive test coverage. Main concern is the NOT(IS_NULL(expr))
  edge case (#2) which could produce incorrect stats expressions for an unlikely but reachable code path.
```

File: cpp/tests/io/parquet_reader_test.cpp [hunk 25]
```diff
@@ -1403,10 +1403,11 @@ TEST_F(ParquetReaderTest, FilterSimple)
...
   // Filtering AST - table[0] < RAND_MAX/2
-  auto literal_value     = cudf::numeric_scalar<decltype(RAND_MAX)>(RAND_MAX / 2);
-  auto literal           = cudf::ast::literal(literal_value);
-  auto col_ref_0         = cudf::ast::column_reference(0);
-  auto filter_expression = cudf::ast::operation(cudf::ast::ast_operator::LESS, col_ref_0, literal);
+  auto literal_value = cudf::numeric_scalar<decltype(RAND_MAX)>(RAND_MAX / 2);
+  auto literal       = cudf::ast::literal(literal_value);
+  auto col_ref_0     = cudf::ast::column_reference(0);
+  auto filter_expression =
+    cudf::ast::operation(cudf::ast::ast_operator::GREATER, literal, col_ref_0);
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/tests/io/parquet_reader_test.cpp]: Updated comment

File: cpp/src/io/parquet/expression_transform_helpers.hpp [hunk 26]
```diff
@@ -0,0 +1,321 @@
...
+ */
+struct binary_operands {
+  ast::ast_operator op;  ///< Input or inverted operator to normalize the `lit op col` expressions
+  operand_kind lhs_type;
+  operand_kind rhs_type;
+  ast::column_reference const*
+    col_ref;  ///< Reliable only when the expression is of the form `col op lit` or `lit op col`
+  ast::literal const*
+    literal;  ///< Reliable only when the expression is of the form `col op lit` or `lit op col`
+};
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/io/parquet/expression_transform_helpers.hpp:53]: Yeah, that's what the comment says.

File: cpp/src/io/parquet/stats_filter_helpers.cpp [hunk 27]
```diff
@@ -48,36 +48,39 @@ std::reference_wrapper<ast::expression const> stats_columns_collector::visit(
...
+        _has_is_null_operator                      = true;
+      }
+    } else {
+      std::ignore = visit_operands(expr.get_operands());
+    }
+    return expr;
+  }
+
+  // Binary operation
+  auto const [op, lhs_kind, rhs_kind, col_ref, literal] = extract_binary_operands(expr);
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/io/parquet/stats_filter_helpers.cpp]: Not needed

**Comment by mhaseeb123** [comment]
>  2. NOT(IS_NULL(expr)) edge case (stats_filter_helpers.cpp:157-165): When IS_NULL's operand is a sub-expression (not a column), the code pushes NOT(original_IS_NULL_ref) into _stats_expr, referencing the input expression tree rather than a stats-transformed expression. Should degrade to _always_true for NOT(IS_NULL(non_column)).
  
This is already handled since if the operand of the wrapped `IS_NULL` is anything but `column_reference`, it automatically returns an `_always_true` from `visit_operands` which is correctly propagated ahead

File: cpp/src/io/parquet/stats_filter_helpers.cpp [hunk 28]
```diff
@@ -141,21 +141,74 @@ std::reference_wrapper<ast::expression const> stats_expression_converter::visit(
...
+      // Special handling for the NOT operator since is necessary as stats transforms use different
+      // columns (vmin, vmax, is_null) for different operators such that NOT(col < val) is not
+      // equivalent to NOT(vmin < val) and instead is equivalent to vmax >= val.
+      if (input_op == ast_operator::NOT) {
+        auto const* child_operation =
+          dynamic_cast<ast::operation const*>(&expr.get_operands().front().get());
+        if (child_operation != nullptr) {
+          auto const child_op = child_operation->get_operator();
+
+          // If the child operator is IS_NULL, we can safely NULL it without any modifications
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/io/parquet/stats_filter_helpers.cpp]: fixed comment

**Comment by Matt711** [comment]
I checked the cudf-poars unit test and upstream parquet, they still pass 👍🏾

File: cpp/src/io/parquet/expression_transform_helpers.cpp [hunk 29]
```diff
@@ -49,18 +32,45 @@ ast::ast_operator invert_non_commutative_operator(ast::ast_operator op)
...
+  } else if (dynamic_cast<ast::column_name_reference const*>(&operand) != nullptr) {
+    CUDF_UNREACHABLE("Column name references are not supported in Parquet expression transformers");
   } else {
     return operand_kind::EXPRESSION;
   }
 }
 
 }  // namespace
 
+std::optional<ast::ast_operator> transform_operator(ast::ast_operator op, operator_transform mode)
```
Comment (vuule) [inline cpp/src/io/parquet/expression_transform_helpers.cpp]: `mode` could be a template parameter; gets rid of the ternary operator branching at runtime.

Comment (mhaseeb123) [reply cpp/src/io/parquet/expression_transform_helpers.cpp]: Done!
